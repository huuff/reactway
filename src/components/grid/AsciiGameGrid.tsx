
import { Grid } from "../../grid/grid";
import { Fragment } from "react";

type GameGridProps = {
    grid: Grid
} & { className: string };

const AsciiGameGrid = ({ grid, className }: GameGridProps) => (
    <div className={`${className} font-mono leading-none text-lg`}>
        {[...Array(grid.height)].map((_, y) => (
            <Fragment key={`row-${y}`}>
                {
                    [...Array(grid.width)].map((_, x) => (
                        <span key={`${x}-${y}`}>{grid.get(x, y) ? "X" : "O"}</span>
                    ))
                }
                <br key={`row-break-${y}`} />
            </Fragment>
        ))}
    </div>
)

export default AsciiGameGrid;