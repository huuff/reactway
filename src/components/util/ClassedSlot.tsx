import isArray from "lodash/isArray";
import { Children, cloneElement, ReactElement } from "react";

type ClassedSlotProps = {
    children: ReactElement | ReactElement[];
    className: string,
}

function withClassName(element: ReactElement, className: string): ReactElement {
    return cloneElement(
            element,
            { ...element.props, className: `${element.props.className} ${className}`}
        );   
}
const ClassedSlot = ({ children, className }: ClassedSlotProps) => {
    return (
        <>
            {
                isArray(children)
                    ? Children.map(children, (child) => withClassName(child, className))
                    : withClassName(children, className)
            }
        </>
    )
}


export default ClassedSlot;