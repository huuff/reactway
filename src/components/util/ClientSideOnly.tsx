import dynamic from 'next/dynamic';
import React, { ReactNode } from 'react';

type NoSsrProps = {
    children: ReactNode
}

const ClientSideOnly = (props: NoSsrProps) => (
  <React.Fragment>{props.children}</React.Fragment>
);

export default dynamic(() => Promise.resolve(ClientSideOnly), {
  ssr: false
});
