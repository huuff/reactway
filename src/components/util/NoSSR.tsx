import dynamic from 'next/dynamic';
import React, { ReactNode } from 'react';

type NoSsrProps = {
    children: ReactNode
}

const NoSsr = (props: NoSsrProps) => (
  <React.Fragment>{props.children}</React.Fragment>
);

export default dynamic(() => Promise.resolve(NoSsr), {
  ssr: false
});
