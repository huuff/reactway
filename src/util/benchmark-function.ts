type BenchmarkResult<T> = {
    result: T;
    elapsedMs: number;
}

// TODO: Take an options object instead of parameters like logToConsole and name?
const benchmark = <T>(f: () => T, logToConsole = false, name?: string): BenchmarkResult<T> => {
    const startTime = performance.now();
    const result = f();
    const endTime = performance.now();
    const elapsedMs = endTime - startTime;

    logToConsole && console.log(`${name} took ${elapsedMs}ms`);
    return { result, elapsedMs };
}

export { benchmark };