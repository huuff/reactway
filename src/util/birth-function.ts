import seedrandom from "seedrandom";
import { v4 as uuid } from "uuid";

// TODO: In the `util` folder?
function shouldBeBornAlive(seedrandom: seedrandom.PRNG, birthFactor: number): boolean {
    return seedrandom() < birthFactor;
}

function randomSeed(): string {
    return uuid();
}

export { shouldBeBornAlive, randomSeed };