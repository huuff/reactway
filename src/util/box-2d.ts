import tuple from "immutable-tuple";
import { Coordinates } from "../grid/grid";

class Box2D {
    
    constructor(
        public readonly topLeft: Coordinates,
        public readonly bottomRight: Coordinates,
    ) {}

    contains = (coordinates: Coordinates) => {
        const [x, y] = coordinates;
        const [topLeftX, topLeftY] = this.topLeft;
        const [bottomRightX, bottomRightY] = this.bottomRight;

        return (x >= topLeftX && x <= bottomRightX) && (y >= topLeftY && y <= bottomRightY);
    }

    divide = (divisor: number) => new Box2D(
        tuple(Math.floor(this.topLeft[0] / divisor), Math.floor(this.topLeft[1] / divisor)),
        tuple(Math.ceil(this.bottomRight[0] / divisor), Math.ceil(this.bottomRight[1] / divisor)),
    )

}

export { Box2D };