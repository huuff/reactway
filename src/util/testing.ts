import { ConwayStrategy } from "../game/conway-strategy";
import { Coordinates, Grid } from "../grid/grid";
import { coordinatesToString }from "./coordinates-to-string";


function fakeConwayStrategy(nextLiveCells: Coordinates[]): ConwayStrategy {
  return (grid: Grid, coordinates: Coordinates) => {
    const stringCoordinates = coordinatesToString(coordinates);
    return nextLiveCells.some(
      liveCoordinates => stringCoordinates == coordinatesToString(liveCoordinates)
      );
  }
}

export { fakeConwayStrategy };