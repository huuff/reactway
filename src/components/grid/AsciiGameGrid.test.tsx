import AsciiGameGrid from "./AsciiGameGrid";
import { render, screen } from "@testing-library/react";
import { FakeGrid, liveCellsInFakeGrid } from "../../util/testing";
import coordinatesToString from "../../util/coordinates-to-string";

describe("AsciiGameGrid", () => {
    test("live cells are X", () => {
        render(<AsciiGameGrid grid={new FakeGrid(5, 5)} />);

        for (const coord of liveCellsInFakeGrid) {
            expect(screen.getByTestId(coordinatesToString(coord))).toHaveTextContent("X");
        }
    });

    test("dead cells are O", () => {
        render(<AsciiGameGrid grid={new FakeGrid(5, 5)} />);

        expect(screen.getByTestId(coordinatesToString([2, 2]))).toHaveTextContent("O");
    });
})