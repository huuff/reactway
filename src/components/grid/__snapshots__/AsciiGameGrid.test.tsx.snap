// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsciiGameGrid snapshot 1`] = `
<div
  className=" font-mono leading-none text-base"
>
  <span
    className="mx-1 hover:bg-hovered-light-dead-cell text-light-text-color"
    data-testid="(0,0)"
    onClick={[Function]}
  >
    O
  </span>
  <span
    className="mx-1 hover:bg-hovered-light-dead-cell text-light-text-color"
    data-testid="(1,0)"
    onClick={[Function]}
  >
    O
  </span>
  <span
    className="mx-1 hover:bg-hovered-light-dead-cell text-light-text-color"
    data-testid="(2,0)"
    onClick={[Function]}
  >
    O
  </span>
  <span
    className="mx-1 hover:bg-hovered-light-dead-cell text-light-text-color"
    data-testid="(3,0)"
    onClick={[Function]}
  >
    O
  </span>
  <span
    className="mx-1 hover:bg-hovered-light-dead-cell text-light-text-color"
    data-testid="(4,0)"
    onClick={[Function]}
  >
    O
  </span>
  <br />
  <span
    className="mx-1 hover:bg-hovered-light-dead-cell text-light-text-color"
    data-testid="(0,1)"
    onClick={[Function]}
  >
    O
  </span>
  <span
    className="mx-1 hover:bg-hovered-light-dead-cell text-light-text-color"
    data-testid="(1,1)"
    onClick={[Function]}
  >
    O
  </span>
  <span
    className="mx-1 hover:bg-hovered-light-alive-cell text-light-text-color"
    data-testid="(2,1)"
    onClick={[Function]}
  >
    X
  </span>
  <span
    className="mx-1 hover:bg-hovered-light-dead-cell text-light-text-color"
    data-testid="(3,1)"
    onClick={[Function]}
  >
    O
  </span>
  <span
    className="mx-1 hover:bg-hovered-light-dead-cell text-light-text-color"
    data-testid="(4,1)"
    onClick={[Function]}
  >
    O
  </span>
  <br />
  <span
    className="mx-1 hover:bg-hovered-light-dead-cell text-light-text-color"
    data-testid="(0,2)"
    onClick={[Function]}
  >
    O
  </span>
  <span
    className="mx-1 hover:bg-hovered-light-dead-cell text-light-text-color"
    data-testid="(1,2)"
    onClick={[Function]}
  >
    O
  </span>
  <span
    className="mx-1 hover:bg-hovered-light-dead-cell text-light-text-color"
    data-testid="(2,2)"
    onClick={[Function]}
  >
    O
  </span>
  <span
    className="mx-1 hover:bg-hovered-light-alive-cell text-light-text-color"
    data-testid="(3,2)"
    onClick={[Function]}
  >
    X
  </span>
  <span
    className="mx-1 hover:bg-hovered-light-alive-cell text-light-text-color"
    data-testid="(4,2)"
    onClick={[Function]}
  >
    X
  </span>
  <br />
  <span
    className="mx-1 hover:bg-hovered-light-dead-cell text-light-text-color"
    data-testid="(0,3)"
    onClick={[Function]}
  >
    O
  </span>
  <span
    className="mx-1 hover:bg-hovered-light-dead-cell text-light-text-color"
    data-testid="(1,3)"
    onClick={[Function]}
  >
    O
  </span>
  <span
    className="mx-1 hover:bg-hovered-light-alive-cell text-light-text-color"
    data-testid="(2,3)"
    onClick={[Function]}
  >
    X
  </span>
  <span
    className="mx-1 hover:bg-hovered-light-alive-cell text-light-text-color"
    data-testid="(3,3)"
    onClick={[Function]}
  >
    X
  </span>
  <span
    className="mx-1 hover:bg-hovered-light-dead-cell text-light-text-color"
    data-testid="(4,3)"
    onClick={[Function]}
  >
    O
  </span>
  <br />
</div>
`;
